File: android/src/main/kotlin/io/sentient/android/settings/systemprompt/SystemPromptViewModel.kt
// ---------------------------------------------------------------------------
// SystemPromptViewModel — System Prompt (Soul.md) page. SLOW save: the soul PUT is
// restart-on-write, so ApplyProfileChangeUseCase(PutSoul) does write + restart in one
// call. Holds `original` (server SoulDoc) + `draft` (editor content); dirty derived.
// Restore-default fetches the canonical template into the draft (no save until Save).
// ---------------------------------------------------------------------------
package io.sentient.android.settings.systemprompt

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import io.sentient.mobiledata.di.SettingsComponent
import io.sentient.mobiledata.result.SentientResult
import io.sentient.mobiledata.usecase.settings.ApplyState
import io.sentient.mobiledata.usecase.settings.ProfileMutation
import io.sentient.mobilesdk.log.createLogger
import io.sentient.mobilesdk.settings.SoulDoc
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

/** System-prompt page state. `original` is the server SoulDoc; `draft` is editor content. */
data class SystemPromptUiState(
    val loading: Boolean = true,
    val loadError: String? = null,
    val original: SoulDoc? = null,
    val draft: String? = null,
    val saving: Boolean = false,
    val restarting: Boolean = false,
    val alreadyApplying: Boolean = false,
    val applyError: String? = null,
) {
    val dirty: Boolean get() = original != null && draft != null && draft != original.content
    val applyActive: Boolean get() = saving || restarting
}

class SystemPromptViewModel(private val component: SettingsComponent) : ViewModel() {
    private val log = createLogger("android", "settings", "system-prompt-vm")

    private val _ui = MutableStateFlow(SystemPromptUiState())
    val ui: StateFlow<SystemPromptUiState> = _ui.asStateFlow()

    init {
        load()
    }

    private fun load() {
        viewModelScope.launch {
            _ui.update { it.copy(loading = true, loadError = null) }
            when (val r = component.profileRepository.getSoul()) {
                is SentientResult.Success ->
                    _ui.update { it.copy(loading = false, original = r.data, draft = r.data.content) }
                is SentientResult.Failure -> {
                    log.warn("load.failed", mapOf("kind" to r.error.kind))
                    _ui.update { it.copy(loading = false, loadError = r.error.userMessage) }
                }
                is SentientResult.Loading -> Unit
            }
        }
    }

    fun setDraft(content: String) {
        _ui.update { it.copy(draft = content, alreadyApplying = false, applyError = null) }
    }

    /** Load the canonical Soul.md into the draft. User still Saves to commit + restart. */
    fun restoreDefault() {
        viewModelScope.launch {
            when (val r = component.profileRepository.getSoulDefault()) {
                is SentientResult.Success -> {
                    log.info("restore-default.loaded")
                    _ui.update { it.copy(draft = r.data.content, applyError = null) }
                }
                is SentientResult.Failure -> {
                    log.warn("restore-default.failed", mapOf("kind" to r.error.kind))
                    _ui.update { it.copy(applyError = r.error.userMessage) }
                }
                is SentientResult.Loading -> Unit
            }
        }
    }

    fun save() {
        val state = _ui.value
        val next = state.draft ?: return
        if (!state.dirty || state.applyActive) return
        log.info("save", mapOf("chars" to next.length))
        viewModelScope.launch {
            component.applyProfileChange(ProfileMutation.PutSoul(next)).collect { st ->
                _ui.update { it.foldApply(st) }
                if (st is ApplyState.Ready) refetch()
            }
        }
    }

    private suspend fun refetch() {
        when (val r = component.profileRepository.getSoul()) {
            is SentientResult.Success -> _ui.update { it.copy(original = r.data, draft = r.data.content) }
            is SentientResult.Failure -> log.warn("refetch.failed", mapOf("kind" to r.error.kind))
            is SentientResult.Loading -> Unit
        }
    }
}

/** Fold one FSM transition into flat progress fields. */
internal fun SystemPromptUiState.foldApply(state: ApplyState): SystemPromptUiState = when (state) {
    ApplyState.Saving -> copy(saving = true, restarting = false, alreadyApplying = false, applyError = null)
    ApplyState.Restarting -> copy(saving = false, restarting = true)
    is ApplyState.Ready -> copy(saving = false, restarting = false, alreadyApplying = false, applyError = null)
    ApplyState.AlreadyApplying -> copy(saving = false, restarting = false, alreadyApplying = true)
    is ApplyState.Failed -> copy(saving = false, restarting = false, applyError = state.error.userMessage)
    ApplyState.Idle -> this
}

File: ios/App/Settings/SystemPrompt/SystemPromptViewModel.swift
// ---------------------------------------------------------------------------
// SystemPromptViewModel — the System Prompt (Soul.md) page state holder. SLOW
// save: PUT /profile/soul is a restart-on-write endpoint, so
// ApplyProfileChangeUseCase.PutSoul blocks through the Hermes worker restart.
//
// Holds `original` (last-saved Soul body) + a Swift-native `draft` string the
// mono editor binds to. Dirty is a plain string diff. "Restore default" fetches
// the canonical template into the draft (NOT saved until the user taps Save).
// ---------------------------------------------------------------------------
import Foundation
import MobileData

@MainActor
@Observable
final class SystemPromptViewModel {
    enum Phase: Equatable {
        case loading
        case ready
        case failed(String)
    }

    enum Save: Equatable {
        case idle
        case saving
        case restarting
        case alreadyApplying
        case applied
        case failed(String)
    }

    private(set) var phase: Phase = .loading
    private(set) var save: Save = .idle

    /// Draft Soul body bound to the mono editor.
    var draft = ""
    /// Whether the restore-default fetch is in flight (blocks the confirm dialog).
    private(set) var isRestoring = false

    private var original = ""
    private let settings: SettingsComponent
    private let log = AppLog("settings", "system-prompt-vm")

    init(settings: SettingsComponent) {
        self.settings = settings
    }

    var isDirty: Bool { phase == .ready && draft != original }
    var isApplying: Bool { save == .saving || save == .restarting }

    func load() async {
        log.info("load")
        do {
            let result = try await settings.profileRepository.getSoul()
            switch onEnum(of: result) {
            case .success(let s):
                original = s.data.content
                draft = s.data.content
                phase = .ready
                log.info("load.ready len=\(draft.count)")
            case .failure(let f):
                phase = .failed(f.error.userMessage)
                log.warn("load.failed kind=\(f.error.kind)")
            case .loading:
                phase = .loading
            }
        } catch is CancellationError {
        } catch {
            phase = .failed("Couldn't load system instructions.")
            log.warn("load.threw")
        }
    }

    /// Fetch the canonical default template into the draft (does not persist).
    func restoreDefault() async {
        log.info("restore-default.start")
        isRestoring = true
        defer { isRestoring = false }
        do {
            let result = try await settings.profileRepository.getSoulDefault()
            switch onEnum(of: result) {
            case .success(let s):
                draft = s.data.content
                log.info("restore-default.applied len=\(draft.count)")
            case .failure(let f):
                save = .failed(f.error.userMessage)
                log.warn("restore-default.failed kind=\(f.error.kind)")
            case .loading:
                break
            }
        } catch is CancellationError {
        } catch {
            save = .failed("Couldn't load the default instructions.")
            log.warn("restore-default.threw")
        }
    }

    func save() async {
        guard isDirty else { return }
        log.info("save.start len=\(draft.count)")
        for await state in settings.applyProfileChange.invoke(mutation: ProfileMutationPutSoul(content: draft)) {
            switch onEnum(of: state) {
            case .idle: break
            case .saving: save = .saving
            case .restarting: save = .restarting
            case .ready:
                log.info("save.ready")
                await load()
                save = .applied
            case .alreadyApplying:
                save = .alreadyApplying
                log.warn("save.already-applying")
            case .failed(let f):
                save = .failed(f.error.userMessage)
                log.warn("save.failed")
            }
        }
    }
}
